### PR TITLE
Kafka script

### PR DESCRIPTION
**Pull request by mehmetalisavas**

### CODE REVIEW

**Comment by mehmetalisavas** [comment]
@cihangir it works just like zookeeper script, if you want i can show you live demo.

File: scripts/consumerproducer/consumer/consumer.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	msgCount int
+)
+
+func main() {
+
+	config := sarama.NewConfig()
+	config.Consumer.Return.Errors = true
+
+	// Specify brokers address. This is default one
+	brokers := []string{"localhost:9092"}
```
Comment (cihangir) [inline scripts/consumerproducer/consumer/consumer.go]: address is not configurable

File: scripts/consumerproducer/consumer/consumer.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	// Get signnal for finish
+	doneCh := make(chan struct{})
+
+	var wg sync.WaitGroup
+	wg.Add(3)
+
+	// warn to return all goroutines when Interrupt signal triggered
+	go handleCtrlC(signals, doneCh)
+
+	for index := 0; index < 3; index++ {
```
Comment (cihangir) [inline scripts/consumerproducer/consumer/consumer.go]: worker count is not configurable

File: scripts/consumerproducer/consumer/consumer.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		select {
+		case err := <-consumer.Errors():
+			fmt.Println(err)
+		case msg := <-consumer.Messages():
+			msgCount++
+			fmt.Println("Received messages", string(msg.Key), string(msg.Value))
+		case <-doneCh:
+			fmt.Println("DONE CHANNEL TRIGGERED")
+			return
+		case <-time.After(10 * time.Second):
```
Comment (cihangir) [inline scripts/consumerproducer/consumer/consumer.go]: deadline is not configurable

File: scripts/consumerproducer/producer/producer.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	// Setup configuration
+	config := sarama.NewConfig()
+	// Return specifies what channels will be populated.
+	// If they are set to true, you must read from
+	// config.Producer.Return.Successes = true
+	// The total number of times to retry sending a message (default 3).
+	config.Producer.Retry.Max = 5
+	// The level of acknowledgement reliability needed from the broker.
+	config.Producer.RequiredAcks = sarama.WaitForAll
+	brokers := []string{"localhost:9092"}
```
Comment (cihangir) [inline scripts/consumerproducer/producer/producer.go]: address is not configurable

File: scripts/consumerproducer/producer/producer.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt)
+
+	var enqueued, keyInt, errors int
+
+	doneCh := make(chan struct{})
+	go func() {
+		for {
+
+			time.Sleep(*interval)
```
Comment (cihangir) [inline scripts/consumerproducer/producer/producer.go]: instead of sleep use time.Ticker

File: scripts/consumerproducer/producer/producer.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+			}
+
+			select {
+			case producer.Input() <- msg:
+				enqueued++
+				fmt.Println("Produce message:", keyInt)
+			case err := <-producer.Errors():
+				errors++
+				fmt.Println("Failed to produce message:", err)
+			case <-signals:
```
Comment (cihangir) [inline scripts/consumerproducer/producer/producer.go]: whenever we want to close, we should not wait for next tick.

File: scripts/consumerproducer/producer/producer.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	go func() {
+		for {
+
+			time.Sleep(*interval)
+
+			strTime := strconv.Itoa(int(time.Now().Unix()))
+			val := fmt.Sprintf("Something Cool: %d", keyInt)
+
+			msg := &sarama.ProducerMessage{
+				Topic: "important",
```
Comment (cihangir) [inline scripts/consumerproducer/producer/producer.go]: topic name should be configurable

File: scripts/consumerproducer/consumer/consumer.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+		panic(err)
+	}
+
+	defer func() {
+		if err := master.Close(); err != nil {
+			panic(err)
+		}
+	}()
+
+	topic := "important"
```
Comment (cihangir) [inline scripts/consumerproducer/consumer/consumer.go]: topic name should be configurable.

File: scripts/generator.go [hunk 9]
```diff
@@ -0,0 +1,173 @@
+package main
```
Comment (cihangir) [inline scripts/generator.go]: why did you implement the same functionality twice?

**Comment by mehmetalisavas** [comment]
@cihangir you can review consumer-producer script. its configurable completely.

File: scripts/generator.go [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+	doneCh := make(chan struct{})
+
+	go func() {
+		defer func() {
+			time.Sleep(*interval / 10)
+			close(doneCh)
+		}()
+
+		for i := 0; i < *element; i++ {
```
Comment (cihangir) [inline scripts/generator.go]: please read my other comments on the deleted files and fix them all...

File: scripts/generator.go [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+// holds the flag variables
+var (
+	element    = flag.Int("element", 10, "default message number")
+	interval   = flag.Duration("interval", 1*time.Second, "default duration to generate strings")
+	goroutines = flag.Int("goroutines", 2, "default go routines count")
+	address    = flag.String("address", "localhost:9092", "addresses for kafka")
+	topic      = flag.String("topic", "defaultTopic", "topic name for kafka")
+)
+
+var mu = &sync.Mutex{}
```
Comment (fatih) [inline scripts/generator.go]: What does this protect?

File: scripts/generator.go [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+	// The total number of times to retry sending a message (default 3).
+	config.Producer.Retry.Max = 5
+
+	// The level of acknowledgement reliability needed from the broker.
+	config.Producer.RequiredAcks = sarama.WaitForAll
+	brokers := strings.Split(*address, ",")
+
+	producer, err := sarama.NewAsyncProducer(brokers, config)
+	if err != nil {
+		panic(err)
```
Comment (fatih) [inline scripts/generator.go]: I would prefer to put all these into separate function. Never do a panic inside main.

File: scripts/generator.go (hunk 11 above)
Comment (mehmetalisavas) [reply scripts/generator.go]: wihle consuming messages, i've counted message. while increasing counter for consumed messages in goroutines, used this protection.

File: scripts/generator.go (hunk 12 above)
Comment (mehmetalisavas) [reply scripts/generator.go]: Actually, consumer and producer were under the different file at the beginning, but i've moved them under the same file. But now, im initializing consumer and producers into seperated function.
thanks for panic advice 👍

File: scripts/generator.go [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+	// CONSUMER GOES HERE
+	config.Consumer.Return.Errors = true
+
+	// Create new consumer
+	master, err := sarama.NewConsumer(brokers, config)
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+
+	consumer, err := master.ConsumePartition(topic, 0, sarama.OffsetNewest)
```
Comment (cihangir) [inline scripts/generator.go]: could you please use the package that we have talked about for consuming cluster wide partitions?

File: scripts/loadgenerator.go [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+	<-c
+	close(cc)
+}
+
+func printUsageErrorAndExit(format string, values ...interface{}) {
+	fmt.Fprintf(os.Stderr, "ERROR: %s\n", fmt.Sprintf(format, values...))
+	fmt.Fprintln(os.Stderr)
+	fmt.Fprintln(os.Stderr, "Available command line options:")
+	flag.PrintDefaults()
+	os.Exit(64)
```
Comment (cihangir) [inline scripts/loadgenerator.go:277]: TIL exit code 64

👍

File: scripts/loadgenerator.go [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+
+	if *verbose {
+		sarama.Logger = logger
+	} else {
+		config.Consumer.Return.Errors = true
+		config.Group.Return.Notifications = true
+	}
+
+	consumer, err := cluster.NewConsumer(brokers, *groupID, strings.Split(*topic, ","), config)
+	// consumer, err := cluster.NewConsumerFromClient(client, *groupID, strings.Split(*topic, ","))
```
Comment (cihangir) [inline scripts/loadgenerator.go:119]: remove this line

**Pull request merged**